### ISSUE TITLE
Schedule Livestreams

### ISSUE THREAD

**Issue by graywolf336** [owncast/owncast#150]
This is a big future thinking request/idea that probably should be split into several or handled as a milestone...or maybe it's too niche and very specific use case that we can dismiss.

I would like to schedule livestreams and be able to link to the individual streams. This would then tie into the recording feature where the stream video endpoint turns into the recorded output, very similar to how YouTube handles it.

---
For an example use case, I will use church services since that has become a big thing with the onslaught of covid-19.

Church A has three services and live streams each of those services. Church A also has an app which takes the stream endpoint and embeds it natively inside the app (iOS and Android) instead of taking them to a web browser or streaming via YouTube. The church wants to have a list of services/events and be able to link to the future and past streams, this way they can schedule it and also have statistics from the app.

To implement this, I think that each livestream event would have its own unique HLS video endpoint. When the event is scheduled in the future, a special graphic would be displayed (or the regular offline could work) with the time on it (they would provide it). Then whenever the stream happens at that time, the video would be there along with the regular HLS endpoint. After the livestream ends, the HLS could be replaced with the recorded video.

Thoughts?

**Comment by gabek** [comment]
Sounds cool, I could totally see this being useful.  I see it as adding the notion of events, and each event has its own metadata and optionally a linked recording.

The biggest challenge I see here is is the web admin UI needed to manage all of it.  It seems like the piece of the puzzle that's required (even more so than recordings, as you could have scheduled events that aren't recorded).  And I'll be honest, I'm bad at building user-friendly forms :). I'd almost rather build a command line tool, but I know that's not not a very good long-term plan.

Going to link to some things here that are already filed that tie into some of this:

Rework of HLS playlist generation (required for recordings) #9
Recordings #102 
Admin dashboard #114

**Comment by gingervitis** [comment]
I like UI. :) I can help with that!

**Comment by gabek** [comment]
Here's a couple of examples of scheduled events.  They both encourage chat before the event starts.  I guess from an engagement standpoint it's a good idea, and because they both have authenticated users they can shut down bad actors.  I always felt no good can come of the chat being available when a stream isn't happening, but that's because we're not identifying people.

YouTube's scheduled event:

![image](https://user-images.githubusercontent.com/414923/94059139-de9d6480-fd96-11ea-955f-8c3dfae8e2e4.png)

[Crowdcast](https://www.crowdcast.io/features) scheduled event:

![image](https://user-images.githubusercontent.com/414923/94059038-bf063c00-fd96-11ea-982c-e4123a4cabdb.png)

**Comment by aaronpk** [comment]
I'd like to bump this up again given the discussion around ActivityPub integration.

Scheduling livestreams is a super useful feature! The main UX benefit in my opinion is that would gives visitors to your OwnCast page an indication that there will be future activity, instead of looking like the stream is just offline. Knowing that something is coming soon gives you a chance to build excitement for that event.

(Of course this is already sort of possible by just editing the description on the home page.)

One implementation note is that with YouTube, as soon as you schedule a livestream it has a permalink already, and that permalink is the same one where the recording is archived to. That could have some implications around the recording feature discussed in #102 as well, for example maybe scheduling a stream creates the recording archive link immediately.

**Comment by gabek** [comment]
I think this would be a cool standalone feature that anybody who's interested could jump on, even without recordings. If anybody is interested in working on a beefy, fun, full feature I'm happy to talk through it with you and plan it out.

I see it as doing the following:

- Be able to create a new scheduled event in the admin that has a date, time, description, maybe a unique image.
- It creates a unique URL where all that metadata is displayed with a countdown.
- The main page lists future events.
- When the event is live then that countdown URL displays like a normal page with the player and chat.

And totally like you said, when recordings happen then you'll be able to watch the recording at that same event URL.

/cc @gingervitis To possibly start thinking about what the UI around this could look like.

**Comment by gabek** [comment]
Some discussion took place here: https://www.youtube.com/watch?v=memUaz03Npg

<blockquote><img src="https://i.ytimg.com/vi/memUaz03Npg/maxresdefault.jpg" width="48" align="right"><div><img src="https://www.youtube.com/s/desktop/f9e10d29/img/favicon_144x144.png" height="14"> YouTube</div><div><strong><a href="https://www.youtube.com/watch?v=memUaz03Npg">Owncast planning and brainstorm series: Scheduled live stream events</a></strong></div><div>Join Nebunez and Gabe as they brainstorm ideas around scheduled live events in Owncast. We talk about what pieces might be required and what functionality th...</div></blockquote>

**Comment by freenandes** [comment]
I've been thinking about this one. Here's what's on my mind. I think we have these concepts to put in the user's mind:
- Schedule
- Event
- Livestream
- and Cowndown UI (to the lack of a better term)

![IMG_20240426_115401a](https://github.com/owncast/owncast/assets/42041153/6d226665-df7a-42e3-8caa-1db739d6812f)

**Comment by freenandes** [comment]
I think this is how it could be structured somehow:
- **Schedule**
*A timeline, potentially a calendar view in the back-office, showing on the public front-end which **Events** will happen on a **Server**.*
    - Belongs to 1 **Server**
    - Belongs to 1 **Countdown**
    - Has zero, 1 or multiple **Events**
    - States
        - Unpublished (default)
        - Published — Shows page X
- **Event**
*An abstract container/object that holds all information about a **Livestream**.*
    - Belongs to 1 **Schedule**
    - Has 1 **Livestream**
    - States
        - Unpublished (default)
        - Published
    - Metadata
        - Streaming date and time
        - State flag
            - Past Event
            - Recent Past Event — The one that is closest to the present
            - Future Event
    - Content
        - Name
        - Description
        - Image
- **Livestream**
*An abstract container/object that holds all information*
    - Belongs to 1 **Event**
    - Type
        - Pre-recorded
            - Video file
            - Show recording (using the video file)
                - Yes
                - No
        - Live stream
        *The **Streamer** will have this URL ready to stream something or manually push the streaming at the scheduled time.*
            - URL
            - URL for recording (optional)
- **Countdown**
*A front-end UI that will be facing the visitor of the **Server** stating what events just went by and what events will come.*
    - Has zero or 1 **Schedule**
    - Has zero or 1 **Recent Past Event**
    - Has zero or 1 **Future Event**
    - States
        - Published — If it has 1 **Published Schedule**
        - Unpublished — If it has
            - 1 **Unpublished Schedule**
            - 1 **Published Schedule** AND zero **Published Event**
    - Content
        - If 1 **Recent Past** **Event**, show this **Event**’s
            - Name
            - Description
            - Calculated time to streaming end
            - Link to recording (if it exists)
        - If 1 **Future Event**, show this **Event**’s
            - Name
            - Description
            - Calculated time to streaming start
    - Actions
        - Has 1 **Future Event**
            - “Get a notification” button for the **Future Event** start
            - “Add to calendar” button
        - Has zero **Future Event** + Has 1 **Past Event**
            - “Get a notification” button for a **Future Event** added to **Schedule**

**Comment by freenandes** [comment]
Given these i think we just need somwhere in the backoffice UI the space to edit the contents of Schedule, Event and Livestream objects.

And also design how the Countdown object will look like.

How does all of this look like to you? Does this structure, so far, fit the problem we're solving here?

**Comment by gabek** [comment]
Upon my first couple looks at it, it looks pretty right to me! A couple small changes.

It's worth calling out we've been talking about an additional state to the page. This shouldn't impact much, but N minutes before a scheduled event is supposed to go live, the chat would become available, even when the stream hasn't started yet. This allows people who show up early to chat a bit before the stream begins. But this new page state shouldn't have major implications.

Also, recordings/replays, won't be a part of this. In the future, this functionality will be added, but that'll be its own thing. So anything about past events probably don't need to be dealt with quite yet.
When stream replays do exist, one idea is that each past event could have a unique URL where you can view the metadata and then watch the replay of it. But we'll cross that bridge when we get there :)

**Comment by freenandes** [comment]
Makes sense. We can further simplify from that point. On the UI i believe we will need:

1. Potentially a full page with a sidebar button for this functionality. It will bring more attention to it. If not, we can think of a more subtle approach on the home page.
2. We will need a list for the **Schedule** object. It will start with an empty state, and a visible button to create **Events**.
3. **Events** will populate the list.
4. These **Events** will have the necessary information for the event to happen. The user can add this with a form. We need to list what we need to support the stream and the chat. URLs? Options for the chat? What are they? Right now, we just know we will need a little info, name, description, and potentially an image upload to animate the event.
5. The **Countdown** will visibly be a simple state on the video player stage. It should at least render all the content the user adds to it.
6. **Events** in the **Schedule** can then be deleted, at a bare minimum. Then should politely provide an option to edit them, in the same creation form. Maybe for the user to be able to write "cancelled" or something like that. If something is not editable, like a date, we can disable that input. Should we provide more functionality to an **Event**? Maybe the ability to change state flag, from published to unpublished? Maybe it's something not needed and assumed auto-published from creation timestamp and unpublished when... well... when deleted!

This is what is coming to mind. Does this make sense?

![IMG_20240429_100155_edit_2047638466972438](https://github.com/owncast/owncast/assets/42041153/bed220b0-4507-4d86-83a2-4414f04dbe08)

**Comment by gabek** [comment]
> Potentially a full page with a sidebar button for this functionality. It will bring more attention to it. If not, we can think of a more subtle approach on the home page.

Just for clarification, you're talking about the admin? I just want to make sure we're talking about the same side of the feature here (admin creation side vs frontend consuming side).

And it's probably already obvious, but in general, as it stands, we'll never have the resources to give the admin the attention to have a really nice user experience. Unfortunate, but true. I just don't want you to spend a lot of time working out the detailed UX of the admin that just can't get any attention. 

I envision most events falling under a reoccurring schedule. "Every Monday at 6:00" kind of thing. We'll, of course, support one-off events as well, but I think a true "schedule" will consist of those. That being said how the date is picked will probably be more complex, since it won't just be a single day. I'm hoping there are some pre-built React calendaring components out there to help us out with regards to this, so maybe we won't have to build this stuff from scratch. So in order to create a schedule of "Every Monday at 6:00" the user will have to be able to select that it happens on Mondays, and then select that it happens at 6:00, and then select that it is reoccurring. And for non-reoccurring events, the user should be able to select a date from a calendar. 

I agree, even though it adds complexity, the user should be able to edit events.

I don't think we should have a "published" state. That starts falling too close into building a CMS with drafts and such, and that's not really what we want to focus on. We want the admin side of this to be as simple as possible, both because it needs to be built, but also just to keep the feature simple to understand.

> The Countdown will visibly be a simple state on the video player stage. It should at least render all the content the user adds to it.

I never thought about the countdown having anything but the actual countdown (1hr 15min until live) being a part of it. But the next scheduled event's name and description should go somewhere. I'm guessing it would replace the offline banner? So that all makes sense that the countdown is actually more than just a countdown.

This feature will probably be iterative over time. For this first version, we don't need an image because there's nowhere to show it. Down the road, when there's a standalone page for an event, then we'll have more flexibility.

It might make sense to go specifically through the pieces of data and how they might be used:

- There's the name of the event that would show in the calendar/schedule itself. For example, "Weekly live stream"
- Owncast has the notion of a "title" that you set when the stream goes live to describe what the stream is about. This should be automatically set when an event happens. I think this would be the same as the description of the event.
- Owncast will automatically send out a reminder (if the user has social features enabled) before this event begins. Should a custom message field be supported? Such as "Don't forget, I'll be doing a special stream today. You won't want to miss it!".

I think that's all that's required for now?

Some components worth looking at for both the admin side and the frontend side (browsing events):

https://github.com/jquense/react-big-calendar
https://scheduler.bitnoise.pl/
https://www.npmjs.com/package/simple-react-calendar


<blockquote><img src="https://opengraph.githubassets.com/281da9467db5e4bb19049669b8e3f2b8515b4d311ebb23e42801ad896b70d49b/jquense/react-big-calendar" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/jquense/react-big-calendar">GitHub - jquense/react-big-calendar: gcal/outlook like calendar component</a></strong></div><div>gcal/outlook like calendar component. Contribute to jquense/react-big-calendar development by creating an account on GitHub.</div></blockquote>
<blockquote><img src="/assets/images/scheduler-hero.png" width="48" align="right"><div><strong><a href="https://scheduler.bitnoise.pl/">React Scheduler</a></strong></div></blockquote>
<blockquote><img src="https://static-production.npmjs.com/338e4905a2684ca96e08c7780fc68412.png" width="48" align="right"><div><img src="https://static-production.npmjs.com/da3ab40fb0861d15c83854c29f5f2962.png" height="14"> npm</div><div><strong><a href="https://www.npmjs.com/package/simple-react-calendar">simple-react-calendar</a></strong></div><div>A simple calendar component for React based applications.. Latest version: 2.2.8, last published: 4 years ago. Start using simple-react-calendar in your project by running `npm i simple-react-calendar`. There are 2 other projects in the npm registry using simple-react-calendar.</div></blockquote>

**Comment by germainelee** [comment]
I've drafted a high level requirements for this feature after watching the brainstorm discussion. 

Do we have a internal knowledge base tool (e.g. Gitbook) to document requirements / technical specifications so that information is consolidated? It would also help to align everyone, including new contributors, towards a common goal and generate great ideas to improve/iterate on this feature over time.

![image](https://github.com/owncast/owncast/assets/118464031/a88dc283-7ea3-4cff-ac60-ec279c68ee90)

**Comment by germainelee** [comment]
Those are great metrics to evaluate open source knowledge base software!

> I ran into [BookStack](https://www.bookstackapp.com/) and [Outline](https://www.getoutline.com/) as options.

Are both of them free to use? I was looking at Outline initially but afraid to mention it because of they charge for Cloud storage. I have no knowledge and experience in self-managed stuff so I'm not sure if it is possible?

**Comment by gabek** [comment]
Oh yeah, I was under the impression we'd be needing to host it ourselves. Except for GitHub wikis. Outline is free if you self-host.

**Comment by germainelee** [comment]
> Oh yeah, I was under the impression we'd be needing to host it ourselves. Except for GitHub wikis. Outline is free if you self-host.

Sorry I'm new to this self-host thing. Does this mean that I would need to set up locally in order to contribute to the docs?

**Comment by gabek** [comment]
Nope! It just means I'd have to set up an instance, so our project could use it. You wouldn't have to do anything.

**Comment by germainelee** [comment]
> Nope! It just means I'd have to set up an instance, so our project could use it. You wouldn't have to do anything.

Oic. If my understanding is correct, that means we create our own Outline accounts and share our email with you to gain access. 

CryptPad offerings look good. It seems like an alternative to office suite (e.g. Google Drive / Microsoft Office), though the document editor doesn't look intuitive (compared to Outline).

I'm leaning towards Outline because its offerings and experience are comparable to the well-known Notion. Below is the top 3 metrics which are critical to me and I've scored them based on how well it satisfies the description of a metric, score ranges from 0 (doesn't satisfy) - 1 (partially satisfy) - 2 (fully satisfy):

**Document import and export** - I can upload and download the page any time.
Outline ([source](https://docs.getoutline.com/s/guide/doc/import-D2ZvLqz411)) : 2
CryptPad ([source](https://docs.cryptpad.org/en/FAQ.html#how-to-import-export-my-documents-to-from-another-platform)) : 1
Bookstack ([source](https://www.bookstackapp.com/docs/user/attachments/)) : 1

**Authentication** - Anyone can easily self-register within the tool with email
Outline ([source](https://docs.getoutline.com/s/guide/doc/authentication-Wr4sfjvmL1)) : 2
CryptPad ([source](https://docs.cryptpad.org/en/user_guide/user_account.html#registration)) : 0
Bookstack ([source](https://www.bookstackapp.com/docs/admin/third-party-auth/)) : 2

**Integration** - I can integrate with existing workflows and tools (e.g. embed Github links)
Outline ([source](https://docs.getoutline.com/s/guide/doc/integrations-eUbAF2WAjF)) : 2
CryptPad ([source](https://docs.cryptpad.org/en/user_guide/apps/code.html)) : 2
Bookstack ([source](https://www.bookstackapp.com/docs/user/attachments/)) : 1

**--- TOTAL SCORE ---**
🥇Outline : 6  
🥉  CryptPad : 3
🥈 Bookstack : 4

**Comment by gabek** [comment]
> CryptPad offerings look good. 

CryptPad is what we had before, and nobody would use it. People just don't like it.

> that means we create our own Outline accounts and share our email with you to gain access.

That's something that would have to be worked out. We could allow anybody to create accounts, but then anybody from the internet can create accounts and create a mess. Or we could lock it down so random people can't create accounts, but that means there's a layer of friction before anybody can even use the tool.

The downside of self-hosting any of these is it's one more thing the project has to maintain. Upgrades, security patches, random outages, etc etc. That's not exactly something I want to be doing. But aside from the GitHub wiki, there are really no other options.

**Comment by gabek** [comment]
Unfortunately Outline requires an auth provider, so it looks like it's not an option for us.

![image](https://github.com/owncast/owncast/assets/414923/dd24a984-3409-4ffb-8345-bf11e77057f9)

**Comment by gabek** [comment]
https://standardnotes.com is another option.

**Comment by gabek** [comment]
We should move this to its own thread. Moving to https://github.com/owncast/owncast/issues/3779

**Comment by germainelee** [comment]
Created a doc in [Huly](https://huly.app/guest/owncast?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NjZiNTVjN2NlOThmOGFhZTlkNWE3M2QiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctZ2FiZWstb3duY2FzdC02NjZhMjI1MC02OWI3MmIzNjI2LTlkYmY3MiIsInByb2R1Y3RJZCI6IiJ9.Jw7ChqxmKSc9u2yfDKv2pLM7s4aj-5412tUHG-x-dto)

**Comment by germainelee** [comment]
Writing it here because not sure what's the best place to put.

What I have in mind is to Display Maximum of 3 Upcoming Events in the Homepage SO THAT:
- Streamer can quickly see my upcoming events without navigating through multiple menus
- Know the existence of this feature after logging in

![image](https://github.com/owncast/owncast/assets/118464031/2e514f1e-b703-4c25-9fe1-8af0a8e8c168)

What are your thoughts?

**Comment by gabek** [comment]
By homepage do you mean the home page of the main web application, or the admin? If you are talking about the admin, that makes sense to me.

**Comment by germainelee** [comment]
> By homepage do you mean the home page of the main web application, or the admin? If you are talking about the admin, that makes sense to me.

Thanks for the prompt reply. yes, the admin.

**Comment by Lislegaard** [comment]
I would be happy with a pretty simple version of this:

- List of scheduled events. Name, time, description and poster would be great but could be minimal.
- Chat automatically activates 15-30 min before scheduled stream for people showing up early.
- Poster image + text saying that "event-name" is starting in x minutes.

**Comment by gabek** [comment]
Yup, those are some of the user-facing features that are planned for scheduling functionality.
